### ISSUE TITLE
Windows users seeing blank "white window" artifacts appear when merging tabs into browser window

### ISSUE THREAD

**Issue by Brave-Matt** [brave/brave-browser#41272]
### Description

![image](https://github.com/user-attachments/assets/98d7a3b1-cd8a-48be-b6b4-35bbc6a88411)

Several users on Windows are seeing strange "white boxes" appear when they tear a tab off from the browser window, then merge it back into that window. It looks like it creates a sort of weird "ghost" window behind the main browser window they're attempting to attach the tab to. A couple users have submitted recordings of the behavior:
https://www.youtube.com/watch?v=GEqN40QRiEA
https://www.youtube.com/watch?v=1dbYKaRoCCE

This appears to only be happening for Windows users as nobody on macOS has reported this issue yet. I cannot reproduce on my end, but both Windows 10 and 11 users have reported the same issue. 

### Steps to reproduce

1. Open a new browser tab.
2. Add a second tab on the same browser window.
3. Drag either of the tabs off the first window.
4. Reattach it to the original browser window.

### Actual result

Strange ghost artifact window appears

### Expected result

No artifacts should appear

### Reproduces how often

Easily reproduced

### Brave version (brave://version info)

v1.70.119

### Channel information

- [X] release (stable)
- [X] beta
- [ ] nightly

### Reproducibility

- [ ] with Brave Shields disabled
- [ ] with Brave Rewards disabled
- [ ] in the latest version of Chrome

### Miscellaneous information

Consolidated user reports:
https://community.brave.com/t/white-windows-when-merging-tabs/570407/11

**Comment by tanoshi40** [comment]
I have the same issue.

2 things to add:
- When splitting and merging windows, after merging when you right-click on a tab select "move to another window", it shows "new window" tabs for each one of the blank windows that seems to be created while merging tabs. When you click on move to that window the "white" window reappears as a proper tab.
- For me on the latest beta build this issue does not seem to be there anymore.

**Comment by Saoiray** [comment]
Just had a user point out that this issue happens by having **Close window when closing last tab** disabled. Upon doing this, I was able to replicate. Upon enabling that setting, the issue goes away. 

Using this same setup, it seems to occur also in Nightly. 

![image](https://github.com/user-attachments/assets/5ad713b9-64c7-40b0-b41a-2ac1e506109a)

### Note:
Chrome doesn't have this setting as an option, so it seems not to be an upstream problem. 

cc: @rebron

**Comment by nightcrewlab** [comment]
Yes, there is no problem when I activate it. I hope this problem will be
solved with the new update.

Thanks

Saoiray ***@***.***>, 28 Eyl 2024 Cmt, 03:20 tarihinde şunu
yazdı:

**Comment by Zoltus** [comment]
I have the exact same problem, and had it for couple of weeks atleast

**Comment by rebron** [comment]
cc: @simonhong

**Comment by Brave-Matt** [comment]
Another user seeing this issue:
https://community.brave.com/t/brave-pages-bug/574139/2

**Comment by Ilie-Lesan** [comment]
Looks like the unload controller is never setting the `is_attempting_to_close_browser_` on true when `kEnableClosingLastTab` is enabled. As a result,  some empty (invisible) windows are created when attaching/detaching a tab from the browser.
STR:
1. Set kEnableClosingLastTab on false
2. Open a browser window with two tabs
3. Drag one tab out of the window
4. Attach the dragged tab to the first window
5. From the context menu, click "Bring all tabs to this window"(shouldn't be visible in this case)

At this point, you will see that the ghost tabs appear in the UI. Also, you can see them in the task manager.
Code of interest:
![Image](https://github.com/user-attachments/assets/f632dc10-c061-46dc-a6d7-3bab0b7df04a)

I didn't do a deep investigation(maybe next week), but I think this is a good starting point if someone else wants to pick it.
FYI: @simonhong @bsclifton

**Comment by Brave-Matt** [comment]
+1 from Community:
https://community.brave.com/t/when-attach-or-detach-a-tab-from-an-active-window-of-brave-this-leave-a-literary-blank-window/574698/2

**Comment by Freefire666** [comment]
Good day
I have now installed Release v1.70.126 portable.
The white windows are still there sporadically and block the desktop :/.

**Comment by Ilie-Lesan** [comment]
@Freefire666 Thanks for the confirmation. The problem is known, and we have steps to reproduce it. I will try to fix it in the near future.

**Comment by q2apro** [comment]
Probably related to: "White window when minimizing a Brave browser window to the system tray in Windows 10" at https://community.brave.com/t/white-window-when-minimizing-a-brave-browser-window-to-the-system-tray-in-windows-10/575265

**Comment by Ilie-Lesan** [comment]
After testing the Brave browser, I saw the following two cases where the ghost windows appear.
1. Trying to close the last window creates two invisible windows. Please see the video below. 

https://github.com/user-attachments/assets/b9694bae-fd76-42b8-bdc6-c560178f8ba3

I couldn't find a clean solution to fix this issue.  It looks like when the  `OnTabStripModelChanged `is called with no tabs is already too late. An attempt to fix this issue was to add a new tab before closing all the tabs scheduled for closing. The problem with this approach is that we cannot solve the problem for reentrancy(Ctrl+W on a tab). Here is how I've initially tried to fix it: https://github.com/Ilie-Lesan/brave-core/commit/eaad7ee17c00d105382ebf2fa7ff73ffda5cad73#diff-c88f3e9d6876f950836824fdc80f905f2095daf7a4f449588afcfe831d297fe5R41

2. Draging in and out a window also creates an invisible(new tab window)

https://github.com/user-attachments/assets/f04557a3-6e7f-4874-b40e-d95fd528a3d8

My initial idea for fixing the second issue(the one with drag & drop) was to allow the browser to close as long as the reason for closing is `kInsertedIntoOtherTabStrip`. This worked well, but the code looked a bit messy(I had to modify the `TabStripEmpty `function to accept a parameter and adapt the implementation for all observers).

The approach from this pull request is to always load a new tab in the existing web content when the user is closing the last tab(inspired by the `PolicyAllowsTabClosing `function). This will solve both problems, but the shortcoming of this approach is that the history of navigation is kept for the new tab. Please let me know if we can live with this behavior or suggest a new approach to loading a new tab.

FYI: @simonhong @bsclifton

**Comment by MadhaviSeelam** [comment]
Verification `In progress` using 
```
Brave | 1.75.133 Chromium: 132.0.6834.57 (Official Build) nightly (64-bit)
-- | --
Revision | ce1787778a7a5d92e63a29f8be965494ef49cbbd
OS | Windows 11 Version 24H2 (Build 26100.2605)
```

Reproduced the issue in 1.70.119 

https://github.com/user-attachments/assets/a47446af-8282-4360-a531-933f46889203
